File: sdks/client-java-core/src/main/java/io/featurehub/client/GoogleAnalyticsCollector.java
package io.featurehub.client;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.net.URLEncoder;
import java.nio.charset.StandardCharsets;
import java.util.List;
import java.util.Map;

public class GoogleAnalyticsCollector implements AnalyticsCollector {
  private static final Logger log = LoggerFactory.getLogger(GoogleAnalyticsCollector.class);
  private final String uaKey; // this must be provided
  private String cid; // if this is null, we will look for it in "other" and log an error if it isn't there
  private final GoogleAnalyticsApiClient client;

  public GoogleAnalyticsCollector(String uaKey, String cid, GoogleAnalyticsApiClient client) {
    this.uaKey = uaKey;
    this.cid = cid;
    this.client = client;

    if (uaKey == null) {
      throw new RuntimeException("UA id must be provided when using the Google Analytics Collector.");
    }
    if (client == null) {
      throw new RuntimeException("Unable to log any events as there is no client, please configure one.");
    }
  }

  public void setCid(String cid) {
    this.cid = cid;
  }

  @Override
  public void logEvent(String action, Map<String, String> other, List<FeatureStateHolder> featureStateAtCurrentTime) {
    StringBuilder batchData = new StringBuilder();

    String finalCid = cid == null ? other.get("cid") : cid;

    if (finalCid == null) {
      log.error("There is no CID provided for GA, not logging any events.");
      return;
    }

    String ev = (other != null && other.get(GoogleAnalyticsApiClient.GA_VALUE) != null)
      ? ("&ev=" + URLEncoder.encode(other.get(GoogleAnalyticsApiClient.GA_VALUE), StandardCharsets.UTF_8)) : "";

    String baseForEachLine = "v=1&tid=" + uaKey + "&cid="+cid+"&t=event&ec=FeatureHub%20Event&ea=" + URLEncoder.encode(action, StandardCharsets.UTF_8) + ev + "&el=";

    featureStateAtCurrentTime.forEach((fsh) -> {

      String line = null;
      if (fsh instanceof FeatureStateBooleanHolder) {
        line = fsh.getBoolean().equals(Boolean.TRUE) ? "on" : "off";
      } else if (fsh instanceof FeatureStateStringHolder) {
        line = fsh.getString();
      } else if (fsh instanceof FeatureStateNumberHolder) {
        line = fsh.getNumber().toPlainString();
      }
      if (line != null) {

        line = URLEncoder.encode(fsh.getKey() + " : " + line, StandardCharsets.UTF_8);
        batchData.append(baseForEachLine);
        batchData.append(line);
        batchData.append("\n");
      }
    });

    if (batchData.length() > 0) {
      client.postBatchUpdate(batchData.toString());
    }

  }
}

File: sdks/client-csharp/FeatureHubSDK/src/GoogleAnalyticsCollector.cs
using System;
using System.Collections.Generic;
using System.Web;
using Common.Logging;
using IO.FeatureHub.SSE.Model;

namespace FeatureHubSDK
{
  public static class GoogleConstants
  {
    // if you wish to pass in the "value" field to analytics, add this to the "other" dictionary
    public const string GaValue = "gaValue";
    // if you wish to override the CID, add this to the "other" dictionary
    public const string Cid = "cid";
  }

  public interface IGoogleAnalyticsClient
  {
    void PostBatchUpdate(string batchData);
  }

  public class GoogleAnalyticsCollector : IAnalyticsCollector
  {
    private static readonly ILog log = LogManager.GetLogger<GoogleAnalyticsCollector>();
    private string _uaKey;
    private string _cid;
    private IGoogleAnalyticsClient _client;

    public string Cid
    {
      set
      {
        if (!string.IsNullOrWhiteSpace(value))
        {
          _cid = value;
        }
      }
      get => _cid;
    }

    public GoogleAnalyticsCollector(string uaKey, string cid, IGoogleAnalyticsClient client)
    {
      if (string.IsNullOrWhiteSpace(uaKey))
      {
        throw new ArgumentException("uaKey cannot be null or empty");
      }

      if (client == null)
      {
        throw new ArgumentException("client cannot be null");
      }

      this._uaKey = uaKey;
      this._cid = cid;
      this._client = client;
    }

    public void LogEvent(string action, Dictionary<string, string> other, List<IFeature> featureStates)
    {
      var batchData = "";

      var finalCid = _cid == null ? (other.ContainsKey(GoogleConstants.Cid) ? other[GoogleConstants.Cid] : null) : _cid;

      if (finalCid == null)
      {
        log.Error("There is no CID provided for GA, not logging any events.");
        return;
      }

      var ev = (other != null && other.ContainsKey(GoogleConstants.GaValue))
        ? ("&ev=" + HttpUtility.UrlEncode(other[GoogleConstants.GaValue]))
        : "";

      var baseForEachLine = "v=1&tid=" + _uaKey + "&cid=" + finalCid + "&t=event&ec=FeatureHub%20Event&ea=" +
                            HttpUtility.UrlEncode(action) + ev + "&el=";

      foreach (var fsh in featureStates)
      {
        if (fsh.IsSet)
        {
          string line = null;
          switch (fsh.Type)
          {
            case FeatureValueType.BOOLEAN:
              line = fsh.BooleanValue == true ? "on" : "off";
              break;
            case FeatureValueType.STRING:
              line = fsh.StringValue;
              break;
            case FeatureValueType.NUMBER:
              line = fsh.NumberValue == null ? null : fsh.NumberValue.ToString();
              break;
            case FeatureValueType.JSON:
              break;
          }

          if (line != null)
          {
            line = HttpUtility.UrlEncode(fsh.Key + " : " + line);
            batchData += baseForEachLine + line + "\n";
          }
        }
      }

      if (batchData.Length > 0)
      {
        // Console.WriteLine(batchData);
        _client.PostBatchUpdate(batchData);
      }
    }
  }
}
